File: video-v7-week1/scripts/build-timings.py
#!/usr/bin/env python3
import json
import subprocess
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
AUDIO = ROOT / 'public' / 'audio'
OUT = ROOT / 'src' / 'timings.json'
FPS = 30
SCENES = [f'S{i:02d}' for i in range(1, 14)]
TAIL = {
    'S08': 0.65,
    'S09': 0.45,
    'S12': 1.35,
    'S13': 1.70,
}

def duration(path: Path) -> float:
    proc = subprocess.run([
        'ffprobe', '-v', 'error', '-show_entries', 'format=duration',
        '-of', 'default=noprint_wrappers=1:nokey=1', str(path)
    ], check=True, capture_output=True, text=True)
    return float(proc.stdout.strip())

items = []
for scene in SCENES:
    path = AUDIO / f'{scene}.mp3'
    if not path.exists() or path.stat().st_size < 4096:
        raise SystemExit(f'Missing or invalid audio: {path}')
    seconds = duration(path) + TAIL.get(scene, 0.28)
    items.append({'id': scene, 'seconds': round(seconds, 3)})

data = {'fps': FPS, 'scenes': items}
OUT.write_text(json.dumps(data, indent=2, ensure_ascii=False) + '\n', encoding='utf-8')
print(json.dumps(data, indent=2, ensure_ascii=False))
print(f'TOTAL_SECONDS={sum(i["seconds"] for i in items):.3f}')

File: video-retime-final/scripts/build-timings.mjs
import fs from 'node:fs';
import {execFileSync} from 'node:child_process';
import path from 'node:path';

const root = path.resolve(path.dirname(new URL(import.meta.url).pathname), '..');
const fps = 30;
const tails = {S01:.18,S02:.18,S03:.18,S04:.22,S05:.18,S06:.2,S07:.25,S08:.55,S09:.35,S10:.35,S11:.3,S12:1.15,S13:1.5};
const scenes = [];
for (let i = 1; i <= 13; i++) {
  const id = `S${String(i).padStart(2,'0')}`;
  const file = path.join(root, 'public', 'audio', `${id}.mp3`);
  const out = execFileSync('ffprobe', ['-v','error','-show_entries','format=duration','-of','default=nw=1:nk=1',file], {encoding:'utf8'}).trim();
  const seconds = Number(out) + (tails[id] ?? .2);
  scenes.push({id, seconds: Number(seconds.toFixed(3))});
}
const data = {fps, scenes};
fs.writeFileSync(path.join(root, 'src', 'timings.json'), JSON.stringify(data, null, 2) + '\n');
console.log(JSON.stringify(data, null, 2));
console.log('TOTAL_SECONDS=' + scenes.reduce((s,x)=>s+x.seconds,0).toFixed(3));
